internal/drivers/azure/nic.go

```diff
@@ -0,0 +1,241 @@
+	"github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/resources/armresources"
+)
+
+const (
+	defaultResourceGroup = "harness-runner-resource"
+)
+
+func (c *config) createResourceGroup(ctx context.Context, cred azcore.TokenCredential) (*armresources.ResourceGroup, error) {
+	logr := logger.FromContext(ctx)
+	if c.resourceGroupName == "" {
+		logr.Debugln("Using default resource group name")
+		c.resourceGroupName = defaultResourceGroup
+	}
+	resourceGroupClient, err := armresources.NewResourceGroupsClient(c.subscriptionID, cred, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	parameters := armresources.ResourceGroup{
+		Location: to.Ptr(c.location),
+	}
+	rg, err := resourceGroupClient.Get(ctx, c.resourceGroupName, nil)
+	if err != nil {
+		resp, createErr := resourceGroupClient.CreateOrUpdate(ctx, c.resourceGroupName, parameters, nil)
+		if createErr != nil {
+			return nil, createErr
+		}
+		logr.Debugf("Created resource group: %s", c.resourceGroupName)
+		return &resp.ResourceGroup, nil
+	}
+	logr.Debugf("Resource group %s already exists", c.resourceGroupName)
+	return &rg.ResourceGroup, nil
+}
+
+func (c *config) createVirtualNetwork(ctx context.Context, cred azcore.TokenCredential, vnetName string) (*armnetwork.VirtualNetwork, error) {
+	logr := logger.FromContext(ctx)
+	vnetClient, err := armnetwork.NewVirtualNetworksClient(c.subscriptionID, cred, nil)
+	if err != nil {
+		return nil, err
+	}
+	parameters := armnetwork.VirtualNetwork{
+		Location: to.Ptr(c.location),
+		Properties: &armnetwork.VirtualNetworkPropertiesFormat{
+			AddressSpace: &armnetwork.AddressSpace{
+				AddressPrefixes: []*string{
+					to.Ptr("10.1.0.0/16"), // example 10.1.0.0/16
+				},
+			},
+		},
+	}
+
+	pollerResponse, createErr := vnetClient.BeginCreateOrUpdate(ctx, c.resourceGroupName, vnetName, parameters, nil)
+	if createErr != nil {
+		return nil, createErr
+	}
+
+	resp, poolErr := pollerResponse.PollUntilDone(ctx, nil)
+	if poolErr != nil {
+		return nil, poolErr
+	}
+	logr.Debugf("Created vnet: %s", vnetName)
+	return &resp.VirtualNetwork, nil
+}
+
+func (c *config) deleteVirtualNetWork(ctx context.Context, cred azcore.TokenCredential, vnetName string) error {
+	vnetClient, err := armnetwork.NewVirtualNetworksClient(c.subscriptionID, cred, nil)
+	if err != nil {
+		return err
+	}
+
+	pollerResponse, err := vnetClient.BeginDelete(ctx, c.resourceGroupName, vnetName, nil)
+	if err != nil {
+		return err
+	}
+
+	_, err = pollerResponse.PollUntilDone(ctx, nil)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (c *config) createSubnets(ctx context.Context, cred azcore.TokenCredential, subnetName, vnetName string) (*armnetwork.Subnet, error) {
+	logr := logger.FromContext(ctx)
+
+	subnetClient, err := armnetwork.NewSubnetsClient(c.subscriptionID, cred, nil)
+	if err != nil {
+		return nil, err
+	}
+	parameters := armnetwork.Subnet{
+		Properties: &armnetwork.SubnetPropertiesFormat{
+			AddressPrefix: to.Ptr("10.1.10.0/24"),
```

Comment:
consider moving these kinds of strings to top of the file and defining in variables

Reply:
done.